### PR TITLE
[evm] Initial version of the Move-To-Yul cross compiler.

### PR DESCRIPTION
**Pull request by wrwg**
This PR sets up the basic framework for compiling Move to Yul. For a high-level impression what is working so far, check the `*.exp` files in `language/evm/move-to-yul/tests` for the generated Yul.

The design is as follows:

- We use the move-model and the stackless bytecode of the Move prover as an input to the compiler.
- Yul code is generated in module `generator.rs`, with runtime helpers defined in `yul_functions.rs`.
- Callable functions are identified by the `#[callable]` attribute. The compiler generates Yul functions for those callables and all transitively called Move functions. If generic functions are called they are specialized on the way.
- The instruction set for primitive value types is fully implemented, as well is control flow.
- No semantic testing is yet happening. Rather for now we are doing only baseline tests.

There are multiple things missing in this PR. Those can be developed likely in independent workstreams:

- Structs, vectors, and tables, as well as resource memory.
- u256
- Contract dispatcher, including parameter marshalling.
- A better treatment of control flow, discovering the original program structure. (See also TODO in generator.rs around line 320.)
- Unit testing support

Note: there are some baseline file changes in the prover because an added optimization on bytecode level.

## Motivation

Support EVM

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Baseline tests

### CODE REVIEW

**Comment by wrwg** [comment]
Adding a bunch of folks as FYI. We chat about actual reviewers in the standup.

File: language/evm/move-to-yul/src/generator.rs [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+        let fun = &ctx.env.get_function(fun_id.to_qualified_id());
+        // Emit function header
+        let params = (0..fun.get_parameter_count())
+            .map(|idx| ctx.make_local_name(fun, idx))
+            .join(", ");
+        let results = match fun.get_return_count() {
+            0 => "".to_string(),
+            1 => format!(" -> {}", ctx.make_result_name(fun, 0)),
+            _ => format!(
+                " -> ({})",
```
Comment (rahxephon89) [inline language/evm/move-to-yul/src/generator.rs:288]: In Yul, "()" is not needed for returning multiple values

File: language/evm/move-to-yul/src/generator.rs [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+        match bc {
+            Jump(_, l) => {
+                print_loc();
+                emitln!(ctx.writer, "$block := {}", get_block(l))
+            }
+            Branch(_, if_t, if_f, cond) => {
+                print_loc();
+                emitln!(
+                    ctx.writer,
+                    "if {} {{ $block := {} }} else {{ $block := {} }}",
```
Comment (rahxephon89) [inline language/evm/move-to-yul/src/generator.rs:442]: Unfortunately, no "else" block can be defined in Yul. Instead, "switch" is needed: https://docs.soliditylang.org/en/latest/yul.html#if

File: language/evm/move-to-yul/tests/Arithm.exp [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+            function A2_M_g(x) -> $result {
+                let $t1, $t2
+                $t1 := 2
+                /// @src 2:1689:1690
+                $t2 := $AddU64(x, $t1)
+                /// @src 2:1688:1691
+                $result := $t2
+            }
+
+            function A2_M_f(x) -> $result {
```
Comment (rahxephon89) [inline language/evm/move-to-yul/tests/Arithm.exp:376]: There are two A2_M_f(x) in the generated code

File: language/evm/move-to-yul/tests/Arithm.exp [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+                let x, $t1, $t2, $t3
+                $t1 := 0
+                $t2 := 1
+                /// @src 2:2177:2178
+                $t3 := $Sub($t1, $t2)
+                /// @src 2:2175:2180
+                $result := $t3
+            }
+
+            function A2_M_g(x) -> $result {
```
Comment (rahxephon89) [inline language/evm/move-to-yul/tests/Arithm.exp:367]: There are two A2_M_g(x) in the generated code

File: language/evm/move-to-yul/src/generator.rs (hunk 1 above)
Comment (wrwg) [reply language/evm/move-to-yul/src/generator.rs]: Done

File: language/evm/move-to-yul/src/generator.rs (hunk 2 above)
Comment (wrwg) [reply language/evm/move-to-yul/src/generator.rs]: Done

File: language/evm/move-to-yul/tests/Arithm.exp (hunk 4 above)
Comment (wrwg) [reply language/evm/move-to-yul/tests/Arithm.exp]: Done

File: language/evm/move-to-yul/tests/Arithm.exp (hunk 3 above)
Comment (wrwg) [reply language/evm/move-to-yul/tests/Arithm.exp]: Done

File: language/evm/move-to-yul/tests/Arithm.exp [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+                        /// @src 2:1913:1937
+                        if tmp_$1 { $block := 9 } else { $block := 10 }
+                    }
+                    case 9 {
+                        $t8 := 4
+                        $Abort($t8)
+                    }
+                    case 10 {
+                        /// @src 2:1951:1952
+                        $t9 := $LogicalNot(b)
```
Comment (rahxephon89) [inline language/evm/move-to-yul/tests/Arithm.exp:231]: The Solidity compiler complains here: "std::exception::what: Variable b is 1 slot(s) too deep inside the stack". Maybe there are too many local variables on the stack while DUP or SWAP operations can only work until 16

File: language/evm/move-to-yul/src/yul_functions.rs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    res
+}
+
+// The Yul functions supporting the compilation scheme.
+functions! {
+Abort: "(code) {
+    revert(0, 0) // TODO: convention to store code
+}",
+AbortBuiltin: "() {
+    $Abort(-1)
```
Comment (rahxephon89) [inline language/evm/move-to-yul/src/yul_functions.rs:83]: Solidity compiler complains about "-1", maybe use 0 or 1 instead?

File: language/evm/move-to-yul/tests/MoveCalls.exp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                $result := x
+            }
+
+            function A2_M_g(x) -> ($result0, $result1) {
+                /// @src 2:309:315
+                $result0 := x
+                $result1 := x
+            }
+
+            function $AbortBuiltin() {
```
Comment (rahxephon89) [inline language/evm/move-to-yul/tests/MoveCalls.exp:46]: $Abort is not generated here?

File: language/evm/move-to-yul/tests/Arithm.exp (hunk 5 above)
Comment (wrwg) [reply language/evm/move-to-yul/tests/Arithm.exp:231]: This is unfortunately a serious restriction of Yul. The only workaround seems to put variables into an array on the heap once there are more than a dozen or so.

https://github.com/ethereum/solidity/issues/6928

File: language/evm/move-to-yul/src/yul_functions.rs (hunk 6 above)
Comment (wrwg) [reply language/evm/move-to-yul/src/yul_functions.rs]: Done

File: language/evm/move-to-yul/tests/MoveCalls.exp (hunk 7 above)
Comment (wrwg) [reply language/evm/move-to-yul/tests/MoveCalls.exp:49]: Done

**Review by rahxephon89 — APPROVED** [review]

File: language/evm/move-to-yul/tests/Arithm.move [hunk 8]
```diff
@@ -0,0 +1,257 @@
+// Tests basic arithmetic.
+// TODO: this file is taken from the prover tests. Convert the specs here into test assertions
+//   once unit testing works. Or remove this and use existing Move tests.
+#[contract]
```
Comment (vgao1996) [inline language/evm/move-to-yul/tests/Arithm.move:4]: I wonder if we should add some sort of prefix, like `#[eth_contract]` or `#[eth(contract)]` to clearly indicate that this is an ethereum-specific thing.

File: language/evm/move-to-yul/tests/Arithm.move [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+		let z: u64 = 3;
+		(z, res)
+	}
+	spec add_two_number {
+	    aborts_if x + y > max_u64();
+	    ensures result_1 == 3;
+	    ensures result_2 == x + y;
+	}
+
+    #[callable]
```
Comment (vgao1996) [inline language/evm/move-to-yul/tests/Arithm.move:18]: Same, I wonder if we should make this `#[eth_callable]` or `#[eth(callable)]`

File: language/evm/move-to-yul/tests/Arithm.move (hunk 8 above)
Comment (wrwg) [reply language/evm/move-to-yul/tests/Arithm.move:4]: Good point. We can easily change this if we decide to do so.

**Review by rahxephon89 — DISMISSED** [review]

**Pull request closed without merging**